Allow passing env to GurobiSolver

JackDunnNZ opened the pull request:
This adds the option of passing a pre-existing `Gurobi.Env` as the first arg to `GurobiSolver`. Previously `env` would be created every time the `GurobiMathProgModel` was created, which was causing issues when using Gurobi with floating-use licences if models were being created at a very fast rate. This option allows you to use the recommended setup of a single env for multiple models to get around this ([ref](http://www.gurobi.com/documentation/6.5/refman/c_grbloadenv.html))

The only other change needed is to make sure that `finalize_env` is always false if the user supplied the `env`.

I can confirm that these changes have fixed the issues I was having, and that the tests are passing for me locally modulo #47. Do we need an explicit test for this and if so what should it test?

@mlubin

mlubin commented:
So there's a weird semantics issue here with parameters, since they're shared across a single environment. The following code might do something unexpected:
```jl
e = Gurobi.Env()
s1 = GurobiSolver(e, Method=2, OutputFlag=0)
s2 = GurobiSolver(e, Method=3)
m1 = Model(solver=s1)
# will this use Method = 3??
m2 = Model(solver=s2)
# will this inherit OutputFlag=0?
```

JackDunnNZ commented:
True, good catch. Is there an easy way to reset the environment to its default state, without recreating it, so the parameters are as expected? (I can't see one). 

Another option could be to warn that this will happen and make the user responsible for managing the parameters across different solver objects in a sensible way? The argument being that you should know what you are doing if you need this in the first place

mlubin commented:
According to http://www.gurobi.com/support/faqs:
> Each model gets its own copy of the Gurobi environment; this allows each model to have its own parameter values. However, this means that you must set parameters on the environment for the model, rather than on the master environment. To see how to set parameters on the environment for a model, see the source code in the examples subdirectory, such as the params example.

So we should just set the parameters on the model's copy of the environment.

JackDunnNZ commented:
I think this is already being done (`setparam!` on `Model` uses the
enviroment returned by `GRBgetenv` as per their example, rather than the
global env), but I will add some tests to make sure

On Fri, 8 Jul 2016 at 09:39 Miles Lubin <notifications@github.com> wrote:

> According to http://www.gurobi.com/support/faqs:
>
> Each model gets its own copy of the Gurobi environment; this allows each
> model to have its own parameter values. However, this means that you must
> set parameters on the environment for the model, rather than on the master
> environment. To see how to set parameters on the environment for a model,
> see the source code in the examples subdirectory, such as the params
> example.
>
> So we should just set the parameters on the model's copy of the
> environment.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaOpt/Gurobi.jl/pull/57#issuecomment-231361609>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEf59HqSq-VeVTdMVyKzPqLzBHeTby7pks5qTlMAgaJpZM4JHgb8>
> .
>

mlubin commented:
The ``GurobiMathProgModel`` constructor calls ``setparam!`` on the passed-in environment directly though.

JackDunnNZ commented:
Whoops, okay. Looks like we can move that down a line to after the model is created and call on `m.inner` rather than `env`?

mlubin commented:
Yeah

JackDunnNZ commented:
@mlubin I've made the changes so that the options are set on the model's copy of the env. The only complication is that the model's env is re-created each time in `loadproblem!` and loses all of the old parameters. I fixed it by carrying around the options in the `GurobiMathProgModel` as well so that they could be re-added to the new model's env. Let me know if you can think of a better way.

I also added some simple tests to check that the env is passed through alright, and lives through a finalizer if we supplied it manually. 

The JuMP tests actually cover the other changes, since the model env is repeatedly rebuilt, and would otherwise lose parameters like `QCPDual` and then fail the corresponding tests.

mlubin commented on src/GurobiSolverInterface.jl:
```diff
@@ -14,35 +14,48 @@ type GurobiMathProgModel <: AbstractLinearQuadraticModel
-  m.inner = Model(env, "", finalize_env=true)
+  m.inner = Model(env, "", finalize_env=finalize_env)
+  # Re-set options on new model's env
+  setparam!(m.inner, "InfUnbdInfo", 1)
```
Why is this set here?

mlubin replied:
Oh, forgot this was already being set. Might be worth an in-line comment

JackDunnNZ replied:
Another option would be just to set it on the env itself rather than the model's copy?

mlubin replied:
I'll leave that up to you

mlubin commented:
Some merge conflicts crept in but looks good. Could you document this in the readme as well?

JackDunnNZ commented:
Sure, whereabouts do you want it? At the end of the "Gurobi Environment" section?

mlubin commented:
I would just add a whole new subsection at the bottom. The readme is in bad shape anyway.

JackDunnNZ commented:
Okay, rebased and added a note to the readme (a bit clumsy, feel free to tweak anything). I also pulled out the parameter setting into a function and added a comment there explaining why we set `InfUnbdInfo`. I think it's better to re-set this at the model-level every time since that's what was happening before this change.

mlubin commented:
Thanks!
